src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -1635,49 +1675,123 @@ private void VisitArguments(ImmutableArray<BoundExpression> arguments, Immutable
-                    TrackNullableStateForAssignment(argument, result.Type, result.Slot, parameterType, -1);
-                }
-                if (refKind != RefKind.Out && (object)parameter != null)
+                (ParameterSymbol parameter, TypeSymbolWithAnnotations parameterType) = GetCorrespondingParameter(i, parameters, argsToParamsOpt, expanded);
+                if (parameter is null)
                 {
-                    WarnOnNullReferenceArgument(argument, result.Type, parameter, expanded);
+                    continue;
                 }
+                VisitArgumentWarn(
+                    arguments[i],
+                    GetRefKind(refKindsOpt, i),
+                    parameter,
+                    parameterType,
+                    results[i]);
             }
         }
 
-        private static ParameterSymbol GetCorrespondingParameter(int argumentOrdinal, MethodSymbol method, ImmutableArray<int> argsToParamsOpt, ref bool expanded)
+        private void VisitArgumentWarn(
+            BoundExpression argument,
+            RefKind refKind,
+            ParameterSymbol parameter,
+            TypeSymbolWithAnnotations parameterType,
+            Result result)
         {
-            if ((object)method == null)
+            TypeSymbolWithAnnotations resultType = result.Type;
+            var argumentType = resultType?.TypeSymbol;
+            switch (refKind)
             {
-                expanded = false;
-                return null;
+                case RefKind.None:
+                case RefKind.In:
+                    {
+                        Conversion conversion = GenerateConversion(_conversions, argument, argumentType, parameterType.TypeSymbol);
+                        resultType = InferResultNullability(argument, conversion, parameterType.TypeSymbol, resultType);
+                        if (!ReportNullReferenceArgumentIfNecessary(argument, resultType, parameter, parameterType) &&
+                            !conversion.Exists)
+                        {
+                            ReportNullabilityMismatchInArgument(argument, argumentType, parameter, parameterType.TypeSymbol);
+                        }
+                    }
+                    break;
+                case RefKind.Out:
+                    if (!ReportNullReferenceAssignmentIfNecessary(argument, resultType, parameterType, useLegacyWarnings: UseLegacyWarnings(argument)))
+                    {
+                        HashSet<DiagnosticInfo> useSiteDiagnostics = null;
+                        if (!_conversions.HasIdentityOrImplicitReferenceConversion(parameterType.TypeSymbol, argumentType, ref useSiteDiagnostics))
+                        {
+                            ReportNullabilityMismatchInArgument(argument, argumentType, parameter, parameterType.TypeSymbol);
+                        }
+                    }
+                    TrackNullableStateForAssignment(argument, resultType, result.Slot, parameterType, -1);
```

Comment:
Consider adding comment (reset the nullable state to declared type). Same a few lines below (line 1742).
Consider naming the `-1` (not sure what it means).

Reply:
Added comments and made last parameter optional.

---
In reply to: [182818727](https://github.com/dotnet/roslyn/pull/26107#discussion_r182818727) [](ancestors = 182818727)